packages/http/test/routes.test.ts

```diff
@@ -508,3 +509,128 @@ describe("http: routes", () => {
     });
   });
 });
+
+describe("uri template", () => {
+  async function getOp(code: string) {
+    const ops = await getOperations(code);
+    return ops[0];
+  }
+  describe("extract implicit parameters", () => {
+    async function getParameter(code: string, name: string) {
+      const op = await getOp(code);
+      const param = op.parameters.parameters.find((x) => x.name === name);
+      ok(param);
+      expect(param.name).toEqual(name);
+      return param;
+    }
+
+    function expectPathParameter(param: HttpOperationParameter, expected: PathOptions) {
+      strictEqual(param.type, "path");
+      const { style, explode, allowReserved } = param;
+      expect({ style, explode, allowReserved }).toEqual(expected);
+    }
+
+    it("extract simple path parameter", async () => {
+      const param = await getParameter(`@route("/bar/{foo}") op foo(foo: string): void;`, "foo");
+      expectPathParameter(param, { style: "simple", allowReserved: false, explode: false });
+    });
+
+    it("+ operator map to allowReserved", async () => {
+      const param = await getParameter(`@route("/bar/{+foo}") op foo(foo: string): void;`, "foo");
+      expectPathParameter(param, { style: "simple", allowReserved: true, explode: false });
+    });
+
+    it.each([
+      [";", "matrix"],
+      ["#", "fragment"],
+      [".", "label"],
+      ["/", "path"],
+    ] as const)("%s map to style: %s", async (operator, style) => {
+      const param = await getParameter(
+        `@route("/bar/{${operator}foo}") op foo(foo: string): void;`,
+        "foo"
+      );
+      expectPathParameter(param, { style, allowReserved: false, explode: false });
+    });
+
+    it("extract simple query parameter", async () => {
+      const param = await getParameter(`@route("/bar{?foo}") op foo(foo: string): void;`, "foo");
+      strictEqual(param.type, "query");
+    });
+
+    it("extract simple query continuation parameter", async () => {
+      const param = await getParameter(
+        `@route("/bar?fixed=yes{&foo}") op foo(foo: string): void;`,
+        "foo"
+      );
+      strictEqual(param.type, "query");
+    });
+  });
+
+  describe("build uriTemplate from parameter", () => {
+    it.each([
+      ["@path one: string", "/foo/{one}"],
+      ["@path(#{allowReserved: true}) one: string", "/foo/{+one}"],
+      ["@path(#{explode: true}) one: string", "/foo/{one*}"],
+      [`@path(#{style: "matrix"}) one: string`, "/foo/{;one}"],
+      [`@path(#{style: "label"}) one: string`, "/foo/{.one}"],
+      [`@path(#{style: "fragment"}) one: string`, "/foo/{#one}"],
+      [`@path(#{style: "path"}) one: string`, "/foo/{/one}"],
+      ["@path(#{allowReserved: true, explode: true}) one: string", "/foo/{+one*}"],
```

Comment:
What if the case is not supported by uri tempalte?

Reply:
No this is not supported by uritemplate because
1. its not a real use case, services doing that are outlier, some where revmoed from openapi3 and openapi4 will support uri template so will remove those too
2. The way the design said is we'll do `@encode(ArrayEncoding.pipeJoin)`

Reply:
> openapi4 will support uri template

Could you explain this a little? Because we know that OpenAPI 3 is supporting uri template without adopting its [syntax](https://github.com/OAI/OpenAPI-Specification/issues/778). How about OpenAPI 4?

Reply:
openapi v4 will take uri template syntax directly as the path key. OpenAPI3 support uri template partially.